### PR TITLE
[PFD] Add proper speed behaviour on approach phase and layout changes.

### LINKED ISSUE #304: Managed speed bug should drop to VAPP when APPR phase is activated
When APPR is activated in FMC, the speed bug (magenta triangle) will jump to VAPP if speed mode is managed. However, the autothrust will slow to the next flap speed (S or F speed on PFD) only, and not to the speed bug, until flaps are set for approach. Hence, the behaviour currently is correct but the display of the speed bug is wrong.

### LINKED ISSUE #854: Wrong Speeds managed during approach phase[BUG]
**Mod Version**

Master

**Describe the bug**

The aircraft currently has a managed speed at some random number not sure what it is and should not work this way upon the activation of the approach phase. The speed bug should be at VAPP when in managed however the aircraft should not slow below green dot in clean config. Once flaps 1 is selected aircraft should slow to S speed. Once Flaps 2 & 3 selected aircraft should slow to F speed, Once FLAPS FULL selected Aircraft should then slow to VAPP

**To Reproduce**

1.Fly an approach
2.
3.

**Expected behavior**

When the approach phase is activated the aircraft should slow (IF IN MANAGED MODE) to:
1) Approach phase activated - Slows to green dot speed
2) Flaps 1 Selected - Aircraft should slow to S speed.
3) Flaps 2 and 3 - Aircraft should slow to F speed.
4) Flaps Full - Aircraft slows to VAPP

**Actual behavior**

**References**

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #1091: Managed speed on approach doesn't target VAPP [BUG]
**Mod Version**

Latest

**Describe the bug**

When the approach mode is activated in the perf page of the MCDU, the target speed when pushing for managed speed should be VAPP, or V approach. In the fbw a32nx, the speed bug is incorrectly set to VLS, or V landing speed. 

**To Reproduce**

1. Enter approach conditions in the MCDU approach perf page.
2. Ensure approach phase is activated and push for managed speed.
3. Note the speed bug is at VLS instead of VAPP when configured for landing.

**Expected behavior**

Magenta target speed should be set to VAPP when in managed speed on approach.

**Actual behavior**

Magenta target speed is actually set to VLS.

**References**

Here you can see in the MCDU that VLS is 145 and VAPP is 150, and the magenta speed target is at 145 when it should be at 150 (VAPP).
Screenshot (79)

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #1598: [REQUEST] Managed speed bug position approach phase
**Is your feature request related to a problem? Please describe.**

Currently when activating the approach phase and speed mode is managed, the speed bug moves to GD, S and F speed before to the VApp speed.
image
In real life the managed speed bug goes directly to the VApp speed.
The autothrottle will hold speed to maintain GD, S and F speeds.

**Describe the feature you'd like to see implemented**

When activating approach phase and selecting managed speed mode, the speed bug should be set to VApp speed.
Autothrottle (if engaged) will maintain GD, S and F speed.

**References**

https://www.youtube.com/watch?v=TEi-WX08ov8
image
image

**Additional context**

Discord username (if different from GitHub): DerL30N#3751

### PR DESCRIPTION
**Pull request by MisterChocker**
Fixes #304
Fixes #854
Fixes #1091
Fixes #1598

## Summary of Changes

- When approach phase activated, VApp speed is targeted for managed speed mode.
- Athr will maintain/hold speed at GD, F and S.
- Added V1 indication.
- Added bottom line logic for speedband.
- Added text field for selected speed field above speedband.
- Added text field for selected speed and managed speed below speedband.
- Resized speedband, Upper and Lower line, V-Speeds.
- Changed font sizes.
- Changed line thickness to match right side.
- Changed text field position above speedband (managed speed).
- Moved uppers line 2px up to fix glitch.
- ~~Changed gray background to match photos.~~
- ~~Changed colors to, lime, cyan or light magenta.~~
- ~~Added light magenta.~~
- Adjusted BARO/MDA/DH texts.

## Screenshots (if necessary)

Before:
image
image

After:
image
image
image

## References

image
image
image
image

## Additional context

image

Discord username (if different from GitHub): DerL30N#3751

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by St54Kevin** [comment]
Nice PR. Isn't the speed band like 90kts high? That's hwo you can see if the bug is outside the range

**Comment by MisterChocker** [comment]
> Nice PR. Isn't the speed band like 90kts high? That's hwo you can see if the bug is outside the range

There is a condition and a check for this already, but the text field is not implemented.

**Comment by MisterChocker** [comment]
Hope you guys can enjoy all these eye candy changes while testing.
I know that after takeoff the managed speed will target the flap retraction speed. This is more difficult to change then approach. That's work to be done in the future (someone else or me in a different pr). Haven't yet checked how to solve that.

Yes the speed bug size is a bit off, I don't know how to change these triangles. Anyways someone which name I forgot wanted to  redo all the symbols to match the real ones.

**Comment by MisterChocker** [comment]
Updated colors
image

**Comment by MisterChocker** [comment]
image
image
image

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js [hunk 1]
```diff
@@ -839,10 +884,33 @@ class Jet_PFD_AirspeedIndicator extends HTMLElement {
         }
     }
     updateTargetSpeeds(currentAirspeed) {
+        {
```
Comment (lousybyte) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js:887]: Nitpicking, but any specific reason for the anonymous block?

**Review by lousybyte — APPROVED** [review]
LGTM

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js (hunk 1 above)
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AirspeedIndicator.js:887]: No Idea, I just stick with the original code, they did it the same way

**Comment by danbw9** [comment]
I will leave a proper testing report after my flight but thought I would just remark this quickly. Looks great so far. Just the green neon colour is not consistent across the FMAs and Altitude tape. It looks like only the FMA green text has changed

Screenshot (516)

**Comment by MisterChocker** [comment]
@danbw9 
> I will leave a proper testing report after my flight but thought I would just remark this quickly. Looks great so far. Just the green neon colour is not consistent across the FMAs and Altitude tape. It looks like only the FMA green text has changed

As far as I see the FMAs and the altitude tape are the same green, also for me looking the same in-game.
I double check that.

**Comment by danbw9** [comment]
Name: danbw9#7028
Date of testing: 01/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1604
Object of test: Confirm the changes made to PFD

Testing technique: Load up on RWY. Check the obvious changes. Then Spawn C&D for normal flight

Steps taken:

Check the obvious changes. Start C&D in UKBB, normal start, taxi, take off, CLB, CRZ and APP into UUDD. Take screenshots and compare to references contained within PR and check 15 items that have been added/changed. 

Challenge A/THR with setting different climb speeds in selected speed mode. 

Checked all colours. The magenta looks better. The green text/digits shade looked a bit off but it was probably shadows that made it look different (see screenshot above). 

Did approach and decelerated in managed mode and checked that the VAPP magenta bug speed is already visible on PFD speedtape at Flap 1,2 etc. Check the magenta speed bug does not go to F or S speed anymore

Checked correct location of the various selected and managed speeds.

MDA/DH text looks a lot better now.

Flight notes

Startup: Everything normal

Taxi/Takeoff : Everything normal 

Cruise: Everything normal 

Descent/Landing: Everything normal 

Shutdown: Everything normal 

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: LGTM. Just noticed that when the brightness is high on the PFD DU for example, the lack of contrast between the speed line indicators and the grey background makes the indications more difficult. See screenshots below. However it may be a brightness issue because many parts of the display like the green FD bars on top of the blue part of the PFD. Thank you for your work.

Lower brightness setting
Screenshot (522)

Higher brightness setting
Screenshot (524)

**Comment by MisterChocker** [comment]
@danbw9 I have noticed the brightness too, at daylight everything was nice visible (with default brightness, also increased display brightness) but however in the night I reduced the display brightness by 1 or 2 steps to get good the bloom away.

**Comment by Benjozork** [comment]
Why are you using `--mcduGreen` ? AFAIK, colors are the same on all screens.

**Comment by MisterChocker** [comment]
@Benjozork 
> Why are you using `--mcduGreen` ? AFAIK, colors are the same on all screens.

Because the mcdu and pfd/nd use the same colors for FMAs etc.
This is a screenshot from I pilot explaining why they have the same colors:
> image

**Comment by Benjozork** [comment]
@MisterChocker why does `--displayGreen` not work ? `mcdu` would, by name,, only be for the MCDU.

**Comment by MisterChocker** [comment]
> @MisterChocker why does `--displayGreen` not work ? `mcdu` would, by name,, only be for the MCDU.

the color --displayGreen is defferent from the --mcduGreen. I think display was ment to be for ECAMs which do have the same color but the color looks slightly different because the ECAM displays are a bit different from then the PFD/ND

image

**Comment by lukecologne** [comment]
There is no difference at all between the DUs, they are the same display. The screenshot you show is saying exactly that. The slight difference in screens he is meaning are small differences between different aircraft from wear etc.

**Comment by MisterChocker** [comment]
@lukecologne Agree.
Do we know if the font thickness or text thickness of the ECAM is any different then the of PFD/ND?

I can change the --displayGreen to the current color from --mcduGreen. In fact I have already tried that and the result was okay, but the ECAM looked very aggressive afterwards.

Here a short comparison (First is original, second is witch --mcduGreen)
image
image

**Comment by lukecologne** [comment]
So below is after, and the top is before? It looks fine IMO

**Comment by MisterChocker** [comment]
@lukecologne  That's what it would look like in-game.
image

**Comment by MisterChocker** [comment]
@Benjozork 
I can indeed change `--displayGreen` but this would also effect any other displays, which I did not wanted to happen in this pr since it already covers a ton of changes.
Therefore I have opened another pr, I can close the other pr and include changes made there in this one, if you like.

**Comment by Benjozork** [comment]
Can you instead remove the color changes from here and add them in the other PR ? That would seem more ideal for separating the changes,

**Comment by MisterChocker** [comment]
@Benjozork yes of course, will do that now then update both prs.

**Comment by MisterChocker** [comment]
@Benjozork Alright moved all color changes to the other branch.

**Comment by Benjozork** [comment]
Which should be merged first ?

**Comment by MisterChocker** [comment]
This one because I have added new elements to the PFD. And then I would adopt colors afterwards.

**Comment by MisterChocker** [comment]
I need to fix some stuff, currently not working any more, I broke something.

**Comment by MisterChocker** [comment]
Cannot find the issue right now, all displays are black. I have checked js files, looking good, no syntax errors.
I suspect something odd with the css after the automatic changes.
Will have to look at this later.

I saw automation changed these: (not sure if these could cause any problems)
left side is result from automation, right side is before:
image
image

**Comment by MisterChocker** [comment]
Okay just for the protocol: I have reverted both css files, still dead. So the automation broke nothing.

**Comment by MisterChocker** [comment]
@Benjozork merge b4148d3af86c4f067f30290e0506a5c46762c638 brakes/broke my branch.

**Comment by Benjozork** [comment]
Did you run `npm run build` before testing in-sim ?

**Comment by MisterChocker** [comment]
Only build.py

**Comment by MisterChocker** [comment]
@Benjozork on me haven't run `npm run build`. This branch works with it's latest changes.

All color changes are removed. Features were tested and found working.

**Comment by Benjozork** [comment]
So it works now ?

**Comment by MisterChocker** [comment]
Yes it works, just had to run `npm run build`.

**Comment by Noxleee** [comment]
Name: Noxlee #7727
Date of testing: Current date 03/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1604 
Object of test: PFD, MFD, Flight phases

Testing technique: Test flight EDDF -> EDLP
Steps you've made to test the Pull Request:
1. Started in EDDF with a standard flight plan, no SID or STAR, Just a Transition -> WRB
2. Done the Flight
3. Landed EDLP on ILS 24

Flight notes

Taxi/Takeoff : V1 and Target speed where shown as described

Cruise: Tested different selected speed, speed higher and lower than actual speed tape where visible (also manged speeds where visible) as described

Descent/Landing : DECEL Point was visible on the route, after reaching this point manage speed was aiming for Vapp, so the approach phase was activated. At Flaps 1 Auto thrust holds speed at S. At Flaps 3 auto thrust holds speed at F. At Flaps Full auto thrust holds speed at Vapp.

Issues (If Any): No issues found
Overall Rating: Good
Conclusions of the testing: That worked well and felt good in the approach phase I cannot find any issues, at least on this short flight. All things were working as described!

**Comment by z0rgmeister** [comment]
Name: z0rgmeister#8972
Date of testing: 03/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1604
Object of test: To test proper airspeed behaviour during approach and layout phases
Steps taken: 

1.Circuit with takeoff and full stop landing at RWY27 EGGD with full ILS approach as programmed in the MCDU with INIT B and PERF page calculations.
2.Full Manual Approach with Managed Speed selected on the FCU.
3.Full stop on the runway.

Flight notes: Realtime weather winds 282/27. IAS holding still acurate.

Taxi/Takeoff : Checked V1 showing as per PR.

Climb/Cruise: N/A. Leveled out at 3000 MSL.

Descent/Approach/Landing: Managed speed manual ILS approach with left hand circuit. Manual activation of Approach Phase commanding GD speed. Observed A/THR follow it accurately. Observed A/THR commanding S,F and VAPP speeds in synch with the schedule and when flaps ordered. All speeds held even with variations in pitch attitude. 

Shutdown: NA

Issues: None worth mentioning. Some colour discrepancy in the ILS info block on the PFD, not major.

Overall Rating: Good.

Conclusions of the testing: Excellent job on making the aircraft behave like it should on following the deployment schedule. Small colour difference on the ILS block on the PFD (different magenta tone I guess?) which is not a major issue and not even within the scope of this PR :)

**Comment by MisterChocker** [comment]
> Small color difference on the ILS block on the PFD (different magenta tone I guess?) which is not a major issue and not even within the scope of this PR :)

Whoopsie that might be the new stuff I added and did not set color to match current :o

Nevertheless this will be fixed with #1672

**Comment by Carter633** [comment]
Name: BritishAvgeek YT -- Carter633#1436
Date of testing: 03/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1604
Object of test: Test behaviour of Vapp speeds and layout changes
Steps taken:

1. Full flight from EDDB to EGKK on VATSIM (hence writing this even though it's been merged as I just finished the test)
2. All MCDU info input and approach data set correctly 
3. Flew the flight, vectored through approach and from managed speed into speed restrictions and back again due to ATC but also helped check behaviour

Flight notes: 
Taxi/Takeoff : V speeds through departure were fine, worked as expected

Climb/Cruise: Everything normal.

Descent/Approach/Landing: Managed speed initially, observed speeds being managed properly and as described by the Autothrottle. Checked each stage accordingly and between that had to comply with set speeds too. Aircraft switched normally between the two modes and each time reverting to managed mode, continued to work very well

Shutdown: NA

Issues: None - one small element noted above by z0rgmeister.

Overall Rating: Good.

Conclusions of the testing: great work! Finally having the aircraft manage its speed better through approach phases now with autothrottle behaving properly - thanks for your work!

**Pull request merged**